sql/core/src/main/scala/org/apache/spark/sql/execution/ShuffledRowRDD.scala

```diff
@@ -181,6 +187,9 @@ class ShuffledRowRDD(
 
       case PartialMapperPartitionSpec(mapIndex, _, _) =>
         tracker.getMapLocation(dependency, mapIndex, mapIndex + 1)
+
+      case CoalescedMapperPartitionSpec(startMapIndex, endMapIndex, numReducers) =>
+        tracker.getMapLocation(dependency, startMapIndex, endMapIndex)
```

Comment:
I see this can reduce the partition number, but I'm not sure this approach has benefits of perf. The origin idea of `OptimizeLocalShuffleReader` is make reducer task run at the same executor of target mapper so that it can reduce some network IO.

I think order by partition size only solve the issue partially. If we want to coalesce mappers, shall we check the coalesced mappers are at the same executor or node ?

Reply:
There is a TODO for it: https://github.com/apache/spark/pull/33310/files#diff-9df2adce46395944a9245289a01eb85e797ec460aaef725a8108ec1054e2772aR61

Reply:
This PR has benefits when all the mappers have very small shuffle output and it's better to coalesce them.

Reply:
OK got it